### PR TITLE
User events management

### LINKED ISSUE #396: User Events Management System
This issue is required for the development of the issue on the bot repo https://github.com/python-discord/bot/issues/1172

The system will require two new tables: `User Events` and `Scheduled Events`

### Fields

* Name (primary key)
* Owner (user ID)
* Subscriptions (many-to-many field of users)

### Operations

* Add a new event record
* Add a user to the subs field
* Query subs by name.
* Remove a user from the subs field
* Remove an event

### Fields

* Event name (foreign key)
* Start time (datetime)
* Duration

### Operations

* Add a new schedule for an event
* Query an event's start time by name
* Remove an event's schedule

### PR DESCRIPTION
**Pull request by RohanJnr**
closes #396 

## User Events:
A few changes have been made to the model as there were changes in implementation:

#### The model
- Event name
- Event organizer
- Event description
- Event message ID (discord message ID)

Have a look at the implementation of the event management system on the bot side for details on how these fields are used.
The description of the bot side PR contains the information: https://github.com/python-discord/bot/pull/1269

## Scheduled Events:
A small change has been made to the model design which was mentioned in the issue:
- End time instead of duration field.

This change has been made solely for convenience and ease of development as I find myself re-calculating the end time with the help of start time and duration(even on the bot side).

### CODE REVIEW

File: pydis_site/apps/api/models/bot/scheduled_event.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        unique=True,
+        help_text="The end time of the scheduled user event."
+    )
+
+    def clean(self) -> None:
+        """
+        Check for event time overlap and if an organizer has already scheduled an event.
+
+        Note: not overriding the save() method as the save() method is not called when creating
+        model instances via the admin interface.
```
Comment (ks129) [inline pydis_site/apps/api/models/bot/scheduled_event.py:33]: What is point of this comment? This model doesn't have save function.

**Review by ks129 — COMMENTED** [review]
One comment

File: pydis_site/apps/api/models/bot/scheduled_event.py (hunk 1 above)
Comment (RohanJnr) [reply pydis_site/apps/api/models/bot/scheduled_event.py:33]: The usual approach to perform validation is by overriding the save() method (as DRF using save() method and save() does not call clean() method). Hence, the comment.

Comment (RohanJnr) [reply pydis_site/apps/api/models/bot/scheduled_event.py:33]: Have a look at this
https://github.com/python-discord/site/pull/410/files#diff-205404e5f7cd4c91e669b190befaeff1a67aec62487917fd317557e141735720R394

**Review by ks129 — APPROVED** [review]
Code looks good. Will test functionality after bot side review comments is applied.

**Comment by RohanJnr** [comment]
Closing PR as a new one will be opened for bot#1603

**Pull request closed without merging**